src/main/java/org/jenkinsci/plugins/workflow/cps/CpsThreadDump.java

```diff
@@ -128,7 +128,7 @@ public static CpsThreadDump from(Throwable t) {
     public static CpsThreadDump from(CpsThreadGroup g) {
         // all the threads that share the same head form a logically single thread
         Map<FlowHead, List<CpsThread>> m = new LinkedHashMap<FlowHead,List<CpsThread>>();
-        for (CpsThread t : g.threads.values()) {
+        for (CpsThread t : g.getThreads()) {
```

Comment:
Looks like this is called off of the CPS VM thread in some cases, see `CpsThreadDumpTest.simple`: 

```
java.lang.AssertionError
at org.jenkinsci.plugins.workflow.cps.CpsThreadGroup.assertVmThread(CpsThreadGroup.java:204)
at org.jenkinsci.plugins.workflow.cps.CpsThreadGroup.getThreads(CpsThreadGroup.java:228)
at org.jenkinsci.plugins.workflow.cps.CpsThreadDump.from(CpsThreadDump.java:131)
at org.jenkinsci.plugins.workflow.cps.CpsThreadGroup.getThreadDump(CpsThreadGroup.java:466)
at org.jenkinsci.plugins.workflow.cps.CpsFlowExecution.getThreadDump(CpsFlowExecution.java:1056)
at org.jenkinsci.plugins.workflow.cps.CpsThreadDumpAction.threadDumpSynchronous(CpsThreadDumpAction.java:60)
at org.jenkinsci.plugins.workflow.cps.CpsThreadDumpTest.simple(CpsThreadDumpTest.java:51)
```
I'll investigate more on Monday. I guess we'll need to add some synchronization around this call for safety in case of concurrent modification and make `CpsThreadGroup.getThreads` not `@CpsVmThreadOnly`.

Reply:
The test deliberately bypasses the UI (see note in `threadDumpSynchronous`) so that is not meaningful. At any rate, we do want the UI to display _something_ even if the CPS VM thread is currently blocked on something. (See also [JENKINS-31839](https://issues.jenkins-ci.org/browse/JENKINS-31839).)

Reply:
Now that we are using a `ConcurrentSkipListMap` we shouldn't have any issues even if the VM thread is blocked on a read/write for some reason.